src/plugins/vis_type_timeseries/public/application/components/lib/convert_series_to_datatable.ts

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+import { IndexPattern } from 'src/plugins/data/public';
+import {
+  Datatable,
+  DatatableRow,
+  DatatableColumn,
+  DatatableColumnType,
+} from 'src/plugins/expressions/public';
+import { TimeseriesVisParams } from '../../../types';
+import { PanelData } from '../../../../common/types';
+import { fetchIndexPattern } from '../../../../common/index_patterns_utils';
+import { getDataStart } from '../../../services';
+import { X_ACCESSOR_INDEX } from '../../visualizations/constants';
+
+interface TSVBTables {
+  [key: string]: Datatable;
+}
+
+interface TSVBColumns {
+  id: number;
+  name: string;
+  isSplit: boolean;
+}
+
+export const addMetaToColumns = (
+  columns: TSVBColumns[],
+  indexPattern: IndexPattern,
+  metricsType: string
+): DatatableColumn[] => {
+  return columns.map((column) => {
+    const field = indexPattern.getFieldByName(column.name);
+    const type = (field?.spec.type as DatatableColumnType) || 'number';
+    const cleanedColumn = {
+      id: column.id.toString(),
+      name: column.name,
+      meta: {
+        type,
+        field: column.name,
+        index: indexPattern.title,
+        source: 'esaggs',
+        sourceParams: {
+          enabled: true,
+          indexPatternId: indexPattern?.id,
+          type: type === 'date' ? 'date_histogram' : column.isSplit ? 'terms' : metricsType,
+        },
+      },
+    };
+    return cleanedColumn;
+  });
+};
+
+export const convertSeriesToDataTable = async (
+  model: TimeseriesVisParams,
+  series: PanelData[],
+  initialIndexPattern: IndexPattern
+) => {
+  const tables: TSVBTables = {};
+  for (let layerIdx = 0; layerIdx < model.series.length; layerIdx++) {
+    const layer = model.series[layerIdx];
+    let usedIndexPattern = initialIndexPattern;
+    // The user can overwrite the index pattern of a layer.
+    // In that case, the index pattern should be fetched again.
+    if (layer.override_index_pattern) {
+      const { indexPatterns } = getDataStart();
+      const { indexPattern } = await fetchIndexPattern(layer.series_index_pattern, indexPatterns);
```

Comment:
nit: with multiple series this could be a bottleneck. Probably it is an initial over-engineering step, but I'd may suggest to fetch them before this loop maybe with a `Promise.all` and then loop.
I'd say it as nit, not a blocker as I do not see too many series looping here.

Reply:
I will look into this when implementing the filter click event, but I am pretty sure that this `override_index_pattern ` doesn't happen a lot.